inc/change.class.php

```diff
@@ -145,6 +145,47 @@ function canReopen() {
    }
 
 
+   function prepareInputForAdd($input) {
+      $input =  parent::prepareInputForAdd($input);
+      if ($input === false) {
+         return false;
+      }
+
+      // Manage auto assign
+      $auto_assign_mode = Entity::getUsedConfig('auto_assign_mode', $input['entities_id']);
+
+      switch ($auto_assign_mode) {
+         case Entity::CONFIG_NEVER :
+            break;
+
+         case Entity::AUTO_ASSIGN_HARDWARE_CATEGORY :
+         case Entity::AUTO_ASSIGN_CATEGORY_HARDWARE :
+            // Auto assign tech/group from Category
+            // Changes are not associated to a hardware then both settings behave the same way
+            $input = $this->setTechAndGroupFromItilCategory($input);
+            break;
+      }
+
+      // Replay setting auto assign if set in rules engine or by auto_assign_mode
+      // Do not force status if status has been set by rules
+      if (((isset($input["_users_id_assign"])
+           && ((!is_array($input['_users_id_assign']) &&  $input["_users_id_assign"] > 0)
+               || is_array($input['_users_id_assign']) && count($input['_users_id_assign']) > 0))
+           || (isset($input["_groups_id_assign"])
+           && ((!is_array($input['_groups_id_assign']) && $input["_groups_id_assign"] > 0)
+               || is_array($input['_groups_id_assign']) && count($input['_groups_id_assign']) > 0))
+           || (isset($input["_suppliers_id_assign"])
+           && ((!is_array($input['_suppliers_id_assign']) && $input["_suppliers_id_assign"] > 0)
+               || is_array($input['_suppliers_id_assign']) && count($input['_suppliers_id_assign']) > 0)))
+          && (in_array($input['status'], $this->getNewStatusArray()))
+          && !$this->isStatusComputationBlocked($input)) {
+         $input["status"] = self::ASSIGNED;
+      }
```

Comment:
This code is also found in tickets. several years ago I updated this code in tickets to allow creation of tickets with several actors (the UI allows only one).

If accepted, it would be a good idea to factorize this snippet among ticket, change and problem. See the similar comment in problem.class.php

Reply:
OK for factorization in CommonItilobject.

Obviously, you'll need to take the ticket implementation and test the existence of item

Re-applying `_xxx_id_assign` should be also part of the factorization.
Rule engine and user input must take the priority on entity auto assignment

Reply:
> Re-applying _xxx_id_assign should be also part of the factorization.

Only L166-178 is factorizable. Doing this factorization preserves the rules engine priority thanks to the call to isStatusComputationBlocked()